runners/flink/src/main/java/org/apache/beam/runners/flink/LyftFlinkStreamingPortableTranslations.java

```diff
@@ -192,13 +235,117 @@ public KinesisByteArrayWindowedValueSchema() {
+      implements KinesisDeserializationSchema<WindowedValue<byte[]>> {
+    private static final ObjectMapper mapper = new ObjectMapper();
+    private static final TypeInformation<WindowedValue<byte[]>> ti =
+        new CoderTypeInformation<>(
+            WindowedValue.getFullCoder(ByteArrayCoder.of(), GlobalWindow.Coder.INSTANCE));
+
+    private static String inflate(byte[] deflatedData) throws IOException {
+      Inflater inflater = new Inflater();
+      inflater.setInput(deflatedData);
+
+      // buffer for inflating data
+      byte[] buf = new byte[4096];
+
+      try (ByteArrayOutputStream bos = new ByteArrayOutputStream(deflatedData.length)) {
+        while (!inflater.finished()) {
+          int count = inflater.inflate(buf);
+          bos.write(buf, 0, count);
+        }
+        return bos.toString("UTF-8");
+      } catch (DataFormatException e) {
+        throw new IOException("Failed to expand message", e);
+      }
+    }
+
+    private static long parseDateTime(String datetime) {
+      try {
+        DateTimeFormatter formatterToUse =
+            (datetime.length() - datetime.indexOf('.') == 7)
+                ? DB_DATETIME_FORMATTER
+                : BACKUP_DB_DATETIME_FORMATTER;
+        TemporalAccessor temporalAccessor = formatterToUse.parse(datetime);
+        LocalDateTime localDateTime = LocalDateTime.from(temporalAccessor);
+        ZonedDateTime zonedDateTime = ZonedDateTime.of(localDateTime, GMT);
+        return java.time.Instant.from(zonedDateTime).toEpochMilli();
+      } catch (DateTimeParseException e) {
+        return java.time.Instant.from(ISO_DATETIME_FORMATTER.parse(datetime)).toEpochMilli();
+      }
+    }
 
     @Override
-    public boolean isEndOfStream(WindowedValue<byte[]> nextElement) {
-      return false;
+    public WindowedValue<byte[]> deserialize(
+        byte[] recordValue,
+        String partitionKey,
+        String seqNum,
+        long approxArrivalTimestamp,
+        String stream,
+        String shardId)
+        throws IOException {
+      String inflatedString = inflate(recordValue);
+
+      JsonNode events = mapper.readTree(inflatedString);
+      if (!events.isArray()) {
+        throw new IOException("Events is not an array");
+      }
+      ;
+
+      Iterator<JsonNode> iter = events.elements();
+      long timestamp = Long.MAX_VALUE;
+      while (iter.hasNext()) {
+        JsonNode occurredAt = iter.next().path(EventField.EventOccurredAt.fieldName());
+        try {
+          timestamp = Math.min(parseDateTime(occurredAt.textValue()), timestamp);
+        } catch (DateTimeParseException e) {
+          // skip this timestamp
+        }
+
+        // if we didn't find any valid timestamps, use Long.MIN_VALUE
+        if (timestamp == Long.MAX_VALUE) {
+          timestamp = Long.MIN_VALUE;
+        }
+      }
+
+      return WindowedValue.timestampedValueInGlobalWindow(
+          inflatedString.getBytes(Charset.forName("UTF-8")), new Instant(timestamp));
```

Comment:
I would probably start with just passing on the original payload for now so that we don't change downstream operators while experimenting. Yes, we repeat the decompress, but we also save a little on the transfer to the sdk.

Reply:
I figured this would be an easy time to make this change since we'll have to update pipelines to use the encoding option anyways. I think it's a clear win for pipeline authors, as currently all pipelines will need to start with a base64-decode/inflate step.

 As far as performance, it's hard to say without benchmarking. I'd expect the extra data transfer (which should almost always be to a local python worker) to be offset by not re-decoding/decompressing.

Reply:
My primary concern here is to simplify testing while we are in experimentation mode. Watermark on or off, we will need to compare. Not having to change the decoding on the Python side would be preferable until things are settled.

I don't see a clear win for pipeline authors (rather neutral), the decoding is already in place for existing Kinesis consumers.